pkg/config/validation/validate.go

```diff
@@ -0,0 +1,51 @@
+package config
+
+import (
+	"fmt"
+
+	"github.com/sirupsen/logrus"
+
+	"cuelang.org/go/cue"
+	"cuelang.org/go/cue/build"
+	"cuelang.org/go/encoding/json"
+)
+
+const (
+	olmBundle          = "olmbundle"
+	olmPackage         = "olmpackage"
+	olmChannel         = "olmchannel"
+	olmSkips           = "olmskips"
+	olmSkipRange       = "olmskipRange"
+	olmGVKProvided     = "olmgvkprovided"
+	olmGVKRequired     = "olmgvkrequired"
+	olmPackageProvided = "olmpackageprovided"
+	olmPackageRequired = "olmpackagerequired"
+)
+
+type configValidator struct {
+	runtime  *cue.Runtime
+	instance *build.Instance
+	logger   *logrus.Entry
+}
+
+func (c configValidator) Validate(b []byte, key string) error {
```

Comment:
Can we expect clients to:
- always know the key of the blob they want to validate?
- call this method with every key to check all schema types?

maybe I'm missing something; does `key: ""` check everything?

Reply:
The key is the name of the schema definition (`#something` that you see on the cue file) that you want to validate the JSON blob against. My intention is to let registry to decide the JSON format and let it be responsible for parsing the JSON and "dissect" into pieces.  Then, it will feed pieces into this library to validate them against a specific definition in the default schema or custom schema. I want this part to be configurable based on inputs so carrying the knowledge of config JSON structure into this is not my desire.

Reply:
ah, I see.

I was thinking it would be nice to be able to execute `cue vet my-index-config.json schema.cue` to validate an entire index in one go.

Reply:
Correct, registry needs to have its own so-called schema to parse config. Registry dictates the schema actually. The schema you see under `/manifests` is the default one. The validation command should accept a custom schema if requested and this library will validate against the definitions specified in custom schema instead of the default one.

Even with the `cue vet` command, you will still need to pick a definition in the cue file to validate again the json file. It won't automatically go through the list of definitions and run validation against each of them.

You can still validate top-level `olm.bundle` and `olm.package` blob and the validation will perform the check in all of specs that are nested underneath the top-level such as `properties` and `relatedImages`. However, because the property is not homogeneous AKA we have variety of properties that are different (different fields or different type for the same field) depending on which type it is, the validation is unable to validate individual property type at the top-level. It can only validate the generic schema AKA the fields that are shared between all of properties.